### PR TITLE
PM-11604 Network layer for checking email token, nav to UI if needed.

### PR DESCRIPTION
**Pull request by dseverns-livefront**
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-11604

## 📔 Objective
- Add endpoint to the identity service to check for expired token status
- If error message states the token is expired then mark the result as such
- If we have expired token from email link use root nav state to navigate to the ExpiredLink screen.

## 📸 Screenshots

https://github.com/user-attachments/assets/2f2abb19-87ea-4841-b942-aa42dc135210

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/MainActivity.kt [hunk 1]
```diff
@@ -76,6 +77,13 @@ class MainActivity : AppCompatActivity() {
                     is MainEvent.CompleteAutofill -> handleCompleteAutofill(event)
                     MainEvent.Recreate -> handleRecreate()
                     MainEvent.NavigateToDebugMenu -> navController.navigateToDebugMenuScreen()
+                    is MainEvent.ShowToast -> {
+                        Toast.makeText(
+                            baseContext,
+                            event.message.invoke(resources),
+                            Toast.LENGTH_SHORT,
+                        ).show()
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/MainActivity.kt:85]: Formatting:

```kotlin
Toast
    .makeText(
        baseContext,
        event.message.invoke(resources),
        Toast.LENGTH_SHORT,
    )
    .show()
```

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt [hunk 2]
```diff
@@ -127,4 +129,20 @@ class IdentityServiceImpl(
             .sendVerificationEmail(body = body)
             .map { it?.content }
     }
+
+    override suspend fun verifyEmailRegistrationToken(
+        body: VerifyEmailTokenRequestJson,
+    ): Result<VerifyEmailTokenResponseJson> = unauthenticatedIdentityApi
+        .verifyEmailToken(
+            body = body,
+        ).map {
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:138]: Can we move the `.map {` down a line

File: app/src/main/java/com/x8bit/bitwarden/data/auth/manager/AuthRequestManager.kt [hunk 3]
```diff
@@ -10,7 +10,7 @@ import com.x8bit.bitwarden.data.auth.manager.model.CreateAuthRequestResult
 import kotlinx.coroutines.flow.Flow
 
 /**
- * A manager class for handling authentication fo logging in with remote device.
+ * A manager class for handling authentication for logging in with remote device.
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/manager/AuthRequestManager.kt:13]: 👍

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt [hunk 4]
```diff
@@ -127,4 +129,20 @@ class IdentityServiceImpl(
...
+            VerifyEmailTokenResponseJson.Valid
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:139]: Does it work if you change the return type on the API from `Unit` to `VerifyEmailTokenResponseJson.Valid`?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt [hunk 5]
```diff
@@ -127,4 +129,20 @@ class IdentityServiceImpl(
...
+        }
+        .recoverCatching { throwable ->
+            val bitwardenError = throwable.toBitwardenError()
+            bitwardenError.parseErrorBodyOrNull<VerifyEmailTokenResponseJson.Invalid>(
+                code = 400,
+                json = json,
+            ) ?: throw throwable
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:146]: Can we format this:

```kotlin
val bitwardenError = throwable.toBitwardenError()
bitwardenError
    .parseErrorBodyOrNull<VerifyEmailTokenResponseJson.Invalid>(
        code = 400,
        json = json,
    )
    ?: throw throwable
```

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 6]
```diff
@@ -1256,6 +1259,33 @@ class AuthRepositoryImpl(
...
             )
 
+    override suspend fun validateEmailToken(email: String, token: String): EmailTokenResult {
+        return identityService
+            .verifyEmailRegistrationToken(
+                body = VerifyEmailTokenRequestJson(
+                    email = email,
+                    token = token,
+                ),
+            ).fold(
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1269]: Can we move the `.fold(` down a line

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt (hunk 4 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:139]: I was seeing a serialization exception

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 7]
```diff
@@ -1256,6 +1259,33 @@ class AuthRepositoryImpl(
...
+                onSuccess = {
+                    when (val json = it) {
+                        is VerifyEmailTokenResponseJson.Invalid -> {
+                            if (json.message.contains(other = "expired", ignoreCase = true)) {
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1273]: Could this check be done in the service, and the `VerifyEmailTokenResponseJson` could have a 3rd type of `Expired`.

Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1273]: > I mainly want to hide the check for a specific string.

Pre-edit, I was indifferent, but the rationale of hiding the "ugliness" I am 110% sold.

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt (hunk 4 above)
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:140]: I was facing the same issue it needs to be Unit and then map it afterwards

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt [hunk 8]
```diff
@@ -127,4 +129,35 @@ class IdentityServiceImpl(
...
+
+/**
+ * If the message body contains text related to the token being expired, return
+ * the TokenExpired type. Otherwise, return the original Invalid response.
+ */
+private fun VerifyEmailTokenResponseJson.Invalid.checkForExpiredMessage() =
+    if (message.contains(other = "expired", ignoreCase = true)) {
+        VerifyEmailTokenResponseJson.TokenExpired
+    } else {
+        this
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:162]: Wanna add a test for this

File: app/src/main/java/com/x8bit/bitwarden/MainActivity.kt [hunk 9]
```diff
@@ -76,6 +77,15 @@ class MainActivity : AppCompatActivity() {
                     is MainEvent.CompleteAutofill -> handleCompleteAutofill(event)
                     MainEvent.Recreate -> handleRecreate()
                     MainEvent.NavigateToDebugMenu -> navController.navigateToDebugMenuScreen()
+                    is MainEvent.ShowToast -> {
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/MainActivity.kt:80]: <image src="https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcnUwM2p0cjcwMGR3Mjdvbm1ma3FqdHJsb3JjODFoZnluM2d0bm1zcCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/Lq0inEMgsVfpAsrnS9/giphy.gif" />

File: app/src/main/java/com/x8bit/bitwarden/MainViewModel.kt [hunk 10]
```diff
@@ -295,6 +295,47 @@ class MainViewModel @Inject constructor(
...
+                    )
+                }
+                EmailTokenResult.Expired -> {
+                    specialCircumstanceManager.specialCircumstance = SpecialCircumstance
+                        .RegistrationEvent
+                        .ExpiredRegistrationLink
+                }
+                EmailTokenResult.Success -> {
+                    if (authRepository.activeUserId != null) {
+                        authRepository.hasPendingAccountAddition = true
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/MainViewModel.kt:328]: Should we cover this scenario?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/api/UnauthenticatedIdentityApi.kt [hunk 11]
```diff
@@ -79,4 +80,9 @@ interface UnauthenticatedIdentityApi {
     suspend fun sendVerificationEmail(
         @Body body: SendVerificationEmailRequestJson,
     ): Result<JsonPrimitive?>
+
+    @POST("/accounts/register/verification-email-clicked")
+    suspend fun verifyEmailToken(
+        @Body body: VerifyEmailTokenRequestJson,
+    ): Result<Unit>
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/api/UnauthenticatedIdentityApi.kt:87]: 👍

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt [hunk 12]
```diff
@@ -141,7 +136,8 @@ class RootNavViewModel @Inject constructor(
                     null,
                     -> RootNavState.VaultUnlocked(activeUserId = userState.activeAccount.userId)
 
-                    is SpecialCircumstance.PreLogin.CompleteRegistration -> {
+                    is SpecialCircumstance.RegistrationEvent,
+                    -> {
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt:140]: ⛏️ Can remove the comma and bring the closure up a line.

File: app/src/test/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryTest.kt [hunk 13]
```diff
@@ -5999,6 +6002,89 @@ class AuthRepositoryTest {
...
+
+        assertEquals(
+            EmailTokenResult.Expired,
+            emailTokenResult,
+        )
+    }
+
+    @Suppress("MaxLineLength")
+    @Test
+    fun `validateEmailToken should return error result when service returns error without expired message`() = runTest {
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryTest.kt:6048]: Make sure to run the formatter on these test files

File: app/src/main/java/com/x8bit/bitwarden/data/platform/manager/util/SpecialCircumstanceExtensions.kt
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/manager/util/SpecialCircumstanceExtensions.kt:1]: Should be able to cover these cases pretty easily.

File: app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt [hunk 14]
```diff
@@ -345,14 +355,136 @@ class MainViewModelTest : BaseViewModelTest() {
...
         )
     }
 
+    @Suppress("MaxLineLength")
+    @Test
+    fun `on ReceiveFirstIntent with complete registration data should set the special circumstance to ExpiredRegistration if token is not valid`() {
+        val viewModel = createViewModel()
+        val intentEmail = "email"
+        val token = "token"
+        val completeRegistrationData = mockk<CompleteRegistrationData>() {
```
Comment (SaintPatrck) [inline app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt:372]:
```suggestion
        val completeRegistrationData = mockk<CompleteRegistrationData> {
```

File: app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt [hunk 15]
```diff
@@ -325,9 +332,12 @@ class MainViewModelTest : BaseViewModelTest() {
 
     @Suppress("MaxLineLength")
     @Test
-    fun `on ReceiveFirstIntent with complete registration data should set the special circumstance to CompleteRegistration`() {
+    fun `on ReceiveFirstIntent with complete registration data should set the special circumstance to CompleteRegistration if token is valid`() {
         val viewModel = createViewModel()
-        val completeRegistrationData = mockk<CompleteRegistrationData>()
+        val completeRegistrationData = mockk<CompleteRegistrationData>() {
```
Comment (SaintPatrck) [inline app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt:337]:
```suggestion
        val completeRegistrationData = mockk<CompleteRegistrationData> {
```

File: app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt [hunk 16]
```diff
@@ -345,14 +355,136 @@ class MainViewModelTest : BaseViewModelTest() {
...
+    }
+
+    @Suppress("MaxLineLength")
+    @Test
+    fun `on ReceiveFirstIntent with complete registration data should show toast if token is not valid but unable to determine reason`() =
+        runTest {
+            val viewModel = createViewModel()
+            val intentEmail = "email"
+            val token = "token"
+            val completeRegistrationData = mockk<CompleteRegistrationData>() {
```
Comment (SaintPatrck) [inline app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt:411]:
```suggestion
            val completeRegistrationData = mockk<CompleteRegistrationData> {
```

File: app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt [hunk 17]
```diff
@@ -345,14 +355,136 @@ class MainViewModelTest : BaseViewModelTest() {
...
+        }
+
+    @Suppress("MaxLineLength")
+    @Test
+    fun `on ReceiveFirstIntent with complete registration data should show toast with custom message if token is not valid but unable to determine reason`() =
+        runTest {
+            val viewModel = createViewModel()
+            val intentEmail = "email"
+            val token = "token"
+            val completeRegistrationData = mockk<CompleteRegistrationData>() {
```
Comment (SaintPatrck) [inline app/src/test/java/com/x8bit/bitwarden/MainViewModelTest.kt:452]:
```suggestion
            val completeRegistrationData = mockk<CompleteRegistrationData> {
```

**Review by SaintPatrck — COMMENTED** [review]
Few minor things

**Review by SaintPatrck — APPROVED** [review]

**Pull request merged**